rust_src/src/multibyte.rs

```diff
@@ -88,13 +92,85 @@ impl LispStringRef {
         s.size
     }
 
-    /// Return width of STRING when displayed in the current buffer. Width is
-    /// measured by how many columns it occupies on the screen. When calculating
-    /// width of a multibyte character in STRING, only the base leading-code is
-    /// considered; the validity of the following bytes is not checked.  Tabs in
-    /// STRING are always taken to occupy `tab-width' columns.
+    /// Return width of the string when displayed in the current buffer. The
+    /// width is measured by how many columns it occupies on the screen while
+    /// paying attention to compositions. This is a convenience function for
+    /// `self.display_width(None)`
     pub fn width(self) -> usize {
-        unsafe { lisp_string_width(self.into(), -1, ptr::null_mut(), ptr::null_mut()) as usize }
+        let (width, _) = self.display_width(None);
+        width
+    }
+
+    /// Return width of the string when displayed in the current buffer. The
+    /// width is measured by how many columns it occupies on the screen while
+    /// paying attention to compositions.
+    ///
+    /// With `precision` argument, return the width of longest substring that
+    /// doesn't exceed `precision`, and the number of characters and bytes it
+    /// contains in the returned tuple.
+    pub fn display_width(self, precision: Option<usize>) -> (usize, Option<(usize, usize)>) {
+        // Manually determine if string is unibyte.
+        let len = self.len_chars() as usize;
+        let multibyte = self.len_chars() < self.len_bytes();
+        let distab = unsafe { buffer_display_table() };
+        // Sum width
+        let mut width = 0;
+        // Character index
+        let mut i = 0;
+        // Byte index
+        let mut b = 0;
+
+        while i < len {
+            let (chars, bytes, thiswidth) = unsafe {
```

Comment:
It might be clearer to calculate `(cmp_id, end)` in one block and then assign `(chars, bytes, thiswidth)` in a separate block. Right now the values and the variables are pretty far away from each other, which makes it a little hard to read.

Reply:
That would definitely make things easier to follow.